R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+    add_aggregate_tdm(t0, groups) %>%
+    select(names(tdm_prototype()), all_of(groups))
+}
+
+warn_zero_rows <- function(data) {
+  # NOTE: This function will only work if the allocation method is
+  # portfolio_weight. ownership_weight outputs 0 for all carsten metric values,
+  # and thus we would be dividing by zero otherwise...
+  message <- 'Filtering for "portfolio_weight" allocation, outputs 0 rows'
+  warn(message, class = "has_zero_rows")
+
+  # Pass `data` to inline inside return()
+  invisible(data)
+}
+
+add_time_step <- function(data, t0, t1, t2) {
+  data %>%
+    mutate(
+      time_step = case_when(
+        .data$year == t0 ~ "t0",
+        .data$year == t0 + t1 ~ "plus_t1",
+        .data$year == t0 + t2 ~ "plus_t2"
+      )
+    )
+}
+
+add_monotonic_factor <- function(data, t0, t1, t2, groups) {
+  data <- data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    mutate(
+      end_year = last(.data$year),
+      end_year_is_t0_t2 = .data$end_year == t0 + t2
+    )
+
+  if (all(data$end_year_is_t0_t2)) {
+    monotonic_factors <- data %>%
+      mutate(
+        .is_monotonic = TRUE,
+        monotonic_factor = dplyr::if_else(.data$.is_monotonic, 1, -1),
+        .is_monotonic = NULL
+      ) %>%
+      select(
+        -c(
+          .data$year,
+          .data$end_year,
+          .data$plan_carsten,
+          .data$plan_alloc_wt_tech_prod,
+          .data$scen_alloc_wt_tech_prod,
+          .data$end_year_is_t0_t2
+        )
+      ) %>%
+      distinct() %>%
+      ungroup()
+  } else {
+    monotonic_factors <- data %>%
+      add_time_step(t0, t1, t2) %>%
+      mutate(time_step = case_when(
+        .data$year == .data$end_year ~ "end_year",
+        TRUE ~ .data$time_step
+      )) %>%
+      select(
+        -c(
+          .data$year,
+          .data$end_year,
+          .data$plan_carsten,
+          .data$plan_alloc_wt_tech_prod,
+          .data$end_year_is_t0_t2
+        )
+      ) %>%
+      tidyr::pivot_wider(
+        names_from = .data$time_step,
+        values_from = .data$scen_alloc_wt_tech_prod
+      ) %>%
+      mutate(
+        .increasing_overall = .data$end_year > .data$t0,
+        .increasing_in_interval = .data$plus_t2 > .data$t0,
+        .is_monotonic = .data$.increasing_in_interval & .data$.increasing_overall,
+        monotonic_factor = dplyr::if_else(.data$.is_monotonic, 1, -1),
+        .increasing_overall = NULL,
+        .increasing_in_interval = NULL,
+        .is_monotonic = NULL,
+        t0 = NULL,
+        plus_t1 = NULL,
+        plus_t2 = NULL,
+        end_year = NULL
+      ) %>%
+      ungroup()
+  }
+
+  left_join(data, monotonic_factors, by = groups)
+}
+
+add_tdm <- function(data, groups) {
+  data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    mutate(
+      .numerator = .data$scenario_production_plus_t2 - .data$portfolio_production_plus_t1,
+      .denominator = .data$scenario_production_plus_t2 - .data$scenario_production_t0,
+      # TODO: This case we
```

Comment:
I don't understand this.

Reply:
Ah it seems I forgot to finish my thought here lol. Will fix. 

For my own reference: what I wanted to explain here is that this where we handle the potential for a 0 denominator (resulting in an undefined value). I think we should actually NOT set this to 0, but rather decide on some form of warning, or "TDM not applicable" value to explain why this wouldn't make sense. @Antoine-Lalechere I believe, agreed with warning the user of this. 

I will think on an appropriate solution this week.